mathics/builtin/inference.py

```diff
@@ -0,0 +1,381 @@
+
+    if assumptions_list is None:
+        return remove_nots_when_unnecesary(pred, evaluation).evaluate(evaluation)
+
+    # Expands Logically
+    assumptions_list, cont = logical_expand_assumptions(assumptions_list, evaluation)
+    while cont:
+        assumptions_list, cont = logical_expand_assumptions(
+            assumptions_list, evaluation
+        )
+
+    # Expands algebraically
+    assumptions_list, cont = algebraic_expand_assumptions(assumptions_list, evaluation)
+    while cont:
+        assumptions_list, cont = algebraic_expand_assumptions(
+            assumptions_list, evaluation
+        )
+    assumption_rules = []
+    for pat in assumptions_list:
+        value = True
+        while pat.has_form("Not", 1):
+            value = not value
+            pat = pat._leaves[0]
+
+        if value:
+            symbol_value = SymbolTrue
+            symbol_negate_value = SymbolFalse
+        else:
+            symbol_value = SymbolFalse
+            symbol_negate_value = SymbolTrue
+
+        if pat.has_form("Equal", 2):
+            if value:
+                lhs, rhs = pat._leaves
+                if lhs.is_numeric():
+                    assumption_rules.append(Rule(rhs, lhs))
+                else:
+                    assumption_rules.append(Rule(lhs, rhs))
+            else:
+                assumption_rules.append(Rule(pat, SymbolFalse))
+                symm_pat = Expression(pat._head, pat._leaves[1], pat._leaves[0])
+                assumption_rules.append(Rule(symm_pat, SymbolFalse))
+        elif pat.has_form("Equivalent", 2):
+            assumption_rules.append(Rule(pat, symbol_value))
+            symm_pat = Expression(pat._head, pat._leaves[1], pat._leaves[0])
+            assumption_rules.append(Rule(symm_pat, symbol_value))
+        elif pat.has_form("Less", 2):
+            if value:
+                assumption_rules.append(Rule(pat, SymbolTrue))
+                assumption_rules.append(
+                    Rule(
+                        Expression(pat._head, pat._leaves[1], pat._leaves[0]),
+                        SymbolFalse,
+                    )
+                )
+                for head in ("Equal", "Equivalent"):
+                    assumption_rules.append(
+                        Rule(
+                            Expression(head, pat._leaves[0], pat._leaves[1]),
+                            SymbolFalse,
+                        )
+                    )
+                    assumption_rules.append(
+                        Rule(
+                            Expression(head, pat._leaves[1], pat._leaves[0]),
+                            SymbolFalse,
+                        )
+                    )
+            else:
+                assumption_rules.append(Rule(pat, SymbolFalse))
+        else:
+            assumption_rules.append(Rule(pat, symbol_value))
+    # TODO: expand the pred and assumptions into an standard,
+    # atomized form, and then apply the rules...
+    print("assumptions:")
```

Comment:
Is this print and the one below supposed to be here? 

Change this in the rebase version.